Replace lodash.assign with vanilla JS.

ZebulanStanphill opened the pull request:
## Description
This PR replaces all uses of lodash.assign in the codebase with the standard JS `Object.assign`. The only exception to this are some ES5 documentation examples which I left unchanged since `Object.assign` wasn't introduced until ES6.

This is the first of several PRs I plan to make in order to reduce our usage of `lodash` whenever there is a simple vanilla JS equivalent, in the hopes of removing it as a dependency from several WP packages.

ZebulanStanphill committed:
Replace lodash.assign with vanilla JS.

ZebulanStanphill commented:
(Maybe I should start running the tests before pushing... but then again, they all end up having to get run again anyway, so what's the point? :stuck_out_tongue:)

All good points, @aduth! Yeah, I'm mainly doing this for the sake of 3rd party package usage and speed improvements. I'll make sure to check the performance comparisons. :+1:

I have now switched to an approach using object spread syntax. In my testing, this appears to be even faster than `Object.assign`!

Unfortunately, there are still some test failures: e2e failures this time. But as far as I can tell, they are completely unrelated. Any ideas?

aduth commented:
> I have now switched to an approach using object spread syntax. In my testing, this appears to be even faster than `Object.assign`!

Huh, that's Interesting, and quite unexpected to me. As the one who originally implemented the original code, I remember explicitly doing it this way in mind of the fact that an object spread is effectively the same as:

```js
Object.assign( {}, a, b );
```

...contrasted with as it was chosen to be implemented:

```js
Object.assign( a, b );
```

To me, I would think the fact that assigning the values from sources of both `b` _and `a`_ would make it not as efficient as assigning only the values of `b`.

Then again, since it's all processed by some combination of Babel and core-js (not actually native), it's hard to accurately measure.

ZebulanStanphill commented:
Actually, I wasn't even testing through Babel compilation; I was comparing self-made tests on https://jsbench.me/. Specifically, I was comparing these two:

```js
const settings = {};
settings.attributes = {};
Object.assign(settings.attributes, {thing: {donut: 'cake'}});
```

```js
const settings = {};
settings.attributes = {
  ...settings.attributes,
  thing: {donut: 'cake'}
};
```

I checked, and if `settings.attributes` was set from the start in the first example, it would be faster. But I was trying to simulate the case where `settings.attributes` would be undefined, so you have to set it prior to running `Object.assign`, which apparently turns out to be slower than the second approach.

ZebulanStanphill commented:
After restarting e2e tests several times, the same ones keep failing, but I haven't got a clue as to why. I wonder if my optimizations have revealed a race condition or something... :confused: Any help would be greatly appreciated.

aduth commented:
Is it the same error each time? From the current one, it appears to be one which has been frequently failing lately: #21052.

It might just be bad luck combined with a particularly flakey test. I'll restart the build. Separately, I'd also like to find some time this week to properly debug / fix these current intermittent failures.

ZebulanStanphill committed:
Use faster approach.

ZebulanStanphill commented:
I did some further testing, and I discovered that the second test I made is only faster in Firefox; it is slower in Chrome. I have now discovered an approach that is more verbose, but it is the fastest I've tried on both Chrome and Firefox. It also happens to use very basic JavaScript that probably won't end up using any `core-js` polyfills.

I have rebased this branch and updated it to use this new approach.

The tests that were failing before are no longer failing, but now a different test is failing. :confused:

ZebulanStanphill commented on packages/block-editor/src/hooks/align.js:
```diff
@@ -89,12 +89,21 @@ export function addAttribute( settings ) {
+		// Gracefully handle if settings.attributes is undefined.
+		if (
+			typeof settings.attributes === 'object' &&
+			settings.attributes !== null
```
`typeof x === 'object'` returns `true` when `x` is `null`. There was a proposal to change this for ES6 several years ago, but it was rejected due to not wanting to break existing code. That's why I'm checking to make sure `settings.attributes` does not equal `null` in this if statement.

It might not actually be possible for `settings.attributes` to be `null`, so I might be able to remove the `null` check once we get stricter type checking.

aduth commented:
While I think it's always a good idea to have performance in mind, I also think it's important to balance the benefit in mind of (a) the human time-cost of benchmarking and (b) the maintainability impact of a more complex implementation. And by benefit, we should consider what the real-world impact is. If we're talking about code which will run a dozen times during the average lifecycle of a page session, I don't personally think it's worth getting too caught up in the minutia of ops/sec of various implementations†, especially if there's one option which is more obviously simple.

Which is to say that, between the options:

```js
const alignAttribute = {
    type: 'string',
};

if (
    typeof settings.attributes === 'object' &&
    settings.attributes !== null
) {
    settings.attributes.align = alignAttribute;
} else {
    settings.attributes = {
        align: alignAttribute,
    };
}
```

...and:

```js
settings.attributes = {
    ...settings.attributes,
    align: alignAttribute,
};
```

I'd generally be more in favor of the second, given the circumstances.

_† To be clear, I don't mean to say it's wasted time. I think it can be useful knowledge to keep in mind for any implementation, especially when working in more "hot" code paths than what we're dealing with here._

ZebulanStanphill committed:
Revert to spread approach for better readability.

ZebulanStanphill commented:
Good point; I've reverted to the spread approach.

aduth reviewed (approved):
Looks good 👍
